src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -498,20 +507,27 @@ public void AddDynamicFile(string filePath, ImmutableArray<string> folders)
                 // _filePath of this type (VSProject) points to csproj/vbproj of the project
                 // where filePath points to dynamic file such as cshtml and etc
                 // for now, we use GetAwaiter().GetResult() until this method itself becomes Async
-                var fileInfo = provider.Value.GetDynamicFileInfoAsync(projectId: Id, projectFilePath: _filePath, filePath: filePath, CancellationToken.None).GetAwaiter().GetResult();
+                var fileInfo = _threadingContext.JoinableTaskFactory.Run(() => provider.Value.GetDynamicFileInfoAsync(projectId: Id, projectFilePath: _filePath, filePath: filePath, CancellationToken.None));
```

Comment:
Ugh. Didn't realize that's what you are doing under the covers, I thought you were going to implement a free-threaded implementation. The second we consume this on a background - this is going to cause ThreadPool exhaustion. 

Do we have a bug tracking adding async everywhere? We'll need that before I call this on a background thread.

Reply:
@jasonmalinowski ^

Reply:
@jasonmalinowski ?  

also, added trakcing issue
https://github.com/dotnet/roslyn/issues/30533

Reply:
Jason want the APIs they provided to us to be async since that's the direction these things are moving in - I don't need it and my implementation is free threaded.

Reply:
@davkean @rynowak 
> a free-threaded implementation

this is free thread implementation, I am not sure what you meant by it not free-thread.

I am using JTF.Run here to call async method in a blocking way without blocking UI thread if it ever used on UI thread.

Wait or Result will cause dead lock if this method is called from UI thread, and the async method ever try to switch to UI thread.

let me tag @AArnott to confirm.

Reply:
If this code _ever_ ever calls code that switches UI thread - I cannot call this off the UI thread. I will run out of threads very quickly.

Reply:
@heejaechang Using JTF.Run instead of Task.Wait or Task.Result is indeed a good way to go. But async methods are generally allowed to have thread affinity, so the async work you do inside *may* need the UI thread (now or in the future, directly or transitively). So any form of blocking a thread while async works executes should be avoided if possible to avoid threadpool exhaustion when called from a threadpool thread as @davkean said, and to avoid UI delays if called from the UI thread. And if this code may execute multiple times in parallel, that makes the threadpool exhaustion risk much greater. 
Remember that netbooks may have only *two cores*, so .NET will only allocate 2 thread pool threads on those machines before adding 1 second delays before reluctantly adding each new threadpool thread. So [threadpool exhaustion](https://github.com/Microsoft/vs-threading/blob/master/doc/threadpool_starvation.md) is a very real concern that requires constant vigilance.

Reply:
As this method is internal (the class is internal and implements no interfaces), the [VSTHRD102](https://github.com/Microsoft/vs-threading/blob/master/doc/analyzers/VSTHRD102.md) analyzer should emit a warning here advising you to avoid JTF.Run since you could expose this asynchronously. What prevents you from doing this?

Reply:
long story. changing the existing call paths between 2 teams require some coordination. so it is an intermediate thing which we will remove once we move existing call paths to async, until then, doing this pattern.

right now, callee (implementation) of the async method in JTF.Run do work in a synchronous way, so JTF.Run should be fine here.